File: src/lib/claim-readiness.ts
export type ClaimReadinessFactor = "receipt" | "purchaseDate" | "serialNumber";

export type ClaimReadinessBand = "needs-attention" | "almost-there" | "claim-ready";

export interface ClaimReadinessResult {
  score: number;
  missing: ClaimReadinessFactor[];
  band: ClaimReadinessBand;
}

const WEIGHTS: Record<ClaimReadinessFactor, number> = {
  receipt: 40,
  purchaseDate: 30,
  serialNumber: 30,
};

// Spec 2.9 — deliberately just the three fields Claim Assist (2.5) needs to
// run without stalling. Model number is excluded since every product has
// one by definition; warranty doc/photo/price are excluded since they
// matter elsewhere but aren't blockers to Claim Assist completing.
export function computeClaimReadiness(input: {
  hasReceipt: boolean;
  purchaseDate: string | null;
  serialNumber: string | null;
}): ClaimReadinessResult {
  const missing: ClaimReadinessFactor[] = [];
  let score = 0;

  if (input.hasReceipt) score += WEIGHTS.receipt;
  else missing.push("receipt");

  if (input.purchaseDate) score += WEIGHTS.purchaseDate;
  else missing.push("purchaseDate");

  if (input.serialNumber) score += WEIGHTS.serialNumber;
  else missing.push("serialNumber");

  const band: ClaimReadinessBand = score >= 80 ? "claim-ready" : score >= 40 ? "almost-there" : "needs-attention";

  return { score, missing, band };
}

export const CLAIM_READINESS_LABEL: Record<ClaimReadinessBand, string> = {
  "needs-attention": "Needs attention",
  "almost-there": "Almost there",
  "claim-ready": "Claim ready",
};

File: ios/WarrantyBuddy/Sources/Models/ClaimReadiness.swift
import Foundation

// Direct port of src/lib/claim-readiness.ts — keep these in sync.
enum ClaimReadinessFactor {
    case receipt, purchaseDate, serialNumber
}

enum ClaimReadinessBand: String {
    case needsAttention = "Needs attention"
    case almostThere = "Almost there"
    case claimReady = "Claim ready"
}

struct ClaimReadinessResult {
    let score: Int
    let missing: [ClaimReadinessFactor]
    let band: ClaimReadinessBand
}

enum ClaimReadiness {
    private static let receiptWeight = 40
    private static let purchaseDateWeight = 30
    private static let serialNumberWeight = 30

    static func compute(hasReceipt: Bool, purchaseDate: String?, serialNumber: String?) -> ClaimReadinessResult {
        var missing: [ClaimReadinessFactor] = []
        var score = 0

        if hasReceipt {
            score += receiptWeight
        } else {
            missing.append(.receipt)
        }

        if purchaseDate != nil && !(purchaseDate?.isEmpty ?? true) {
            score += purchaseDateWeight
        } else {
            missing.append(.purchaseDate)
        }

        if serialNumber != nil && !(serialNumber?.isEmpty ?? true) {
            score += serialNumberWeight
        } else {
            missing.append(.serialNumber)
        }

        let band: ClaimReadinessBand
        if score >= 80 {
            band = .claimReady
        } else if score >= 40 {
            band = .almostThere
        } else {
            band = .needsAttention
        }

        return ClaimReadinessResult(score: score, missing: missing, band: band)
    }
}
